systemverilog-plugin/UhdmAst.cc

```diff
@@ -635,10 +617,44 @@ static AST::AstNode *expand_dot(const AST::AstNode *current_struct, const AST::A
         }
         if (c->type == AST::AST_RANGE) {
             // Currently supporting only 1 range
-            log_assert(!struct_range);
+            if (struct_range) {
+                log_error("Currently support for dot-access is limited to single range\n");
+            }
             struct_range = c;
         }
     }
+    AST::AstNode *left = nullptr, *right = nullptr;
+    switch (current_struct_elem->type) {
+    case AST::AST_STRUCT_ITEM:
+        left = AST::AstNode::mkconst_int(current_struct_elem->range_left, true);
+        right = AST::AstNode::mkconst_int(current_struct_elem->range_right, true);
+        break;
+    case AST::AST_STRUCT:
+    case AST::AST_UNION:
+        // TODO(krak): add proper support for accessing struct/union elements
+        // with multirange
+        // Currently support only special access to 2 dimensional packed element
+        // when selecting single range
+        if (struct_range && current_struct_elem->multirange_swapped.size() == 2 /* 2 dimensional */) {
```

Comment:
Wait, shouldn't you take into account `multirange_swapped` values somewhere below? At least in an assert.

Reply:
Oh, I've assumed that size of ``multirange_swapped`` equals to half of size of ``multirange_dimensions`` (and it should always be the case), but you are right, this can be confusing. I've changed this to just ``multirange_dimensions.size() / 2``.

Reply:
It does, I think?
As for readability, I'm not sure this division + assert with modulo is much better. I think `const int dimensions_num = multirange_swapped.size()` and then `dimensions_num == 2` in `if` would be best.

However, what I had in mind in the first comment is:
shouldn't we take into account values stored in `multirange_swapped` when using values like `current_struct_elem->children.front()->range_left` in calculations? I know `multirange_swapped` tells whether ranges are swapped, but I don't know how it affects this use case (if at all).

Reply:
> It does, I think? As for readability, I'm not sure this division + assert with modulo is much better. I think `const int dimensions_num = multirange_swapped.size()` and then `dimensions_num == 2` in `if` would be best.

At first I tried this, but compiler errors-out with message about cross initialization.

> However, what I had in mind in the first comment is: shouldn't we take into account values stored in `multirange_swapped` when using values like `current_struct_elem->children.front()->range_left` in calculations? I know `multirange_swapped` tells whether ranges are swapped, but I don't know how it affects this use case (if at all).

I don't think we need to take into account value stored in ``multirange_swapped`` in the code that I'm changing.

Correct order of ranges (right range lower than left range) is assured by ``add_multirange_attribute`` function and I'm only adjusting size of single element.

``multirange_swapped`` needs to be taken into account when selecting actual element, but I think this is beyond the scope for this PR.

Reply:
Okay, good to know. Thanks.